### PR TITLE
Show hostname first (#1374)

### PR DESCRIPTION
**Pull request by Dalimil**
Associated Issue: #1374 
This was pushed with --no-verify and still needs a clean-up.

out

### CODE REVIEW

File: src/utils/sources-tree.js [hunk 1]
```diff
@@ -242,6 +255,20 @@ function determineFileSortOrder(nodes:Array<Node>, pathPart:string,
...
+      const rootUrlMatch = isExactUrlMatch(pathPart, debuggeeUrl);
+      const nodeUrlMatch = isExactUrlMatch(node.name, debuggeeUrl);
+      if (rootUrlMatch) {
+        // pathPart must go first
+        return true;
+      }
+      if (nodeUrlMatch) {
+        // Examined item must go first
+        return false;
+      }
```
Comment (jasonLaster) [inline src/utils/sources-tree.js]: nit: 

```js
return rootUrlMatch && !nodeUrlMatch
```

File: src/utils/sources-tree.js [hunk 2]
```diff
@@ -285,17 +312,18 @@ function collapseTree(node: any, depth: number = 0) {
...
+    addToTree(uncollapsedTree, source, debuggeeUrl);
   }
   const sourceTree = collapseTree(uncollapsedTree);
 
   return { uncollapsedTree,
     sourceTree,
     parentMap: createParentMap(sourceTree),
-    focusedItem: null };
+    focusedItem: null,
+    debuggeeUrl };
```
Comment (jasonLaster) [inline src/utils/sources-tree.js]: lets get some unit tests in here. Should be pretty similar to what we already have

**Comment by jasonLaster** [comment]
@Dalimil I think this should fix the flow error:

```js
diff --git a/src/actions/types.js b/src/actions/types.js
index 0d3f889..75cf6d6 100755
--- a/src/actions/types.js
+++ b/src/actions/types.js
@@ -149,6 +149,8 @@ type PauseAction =
   | { type: "DELETE_EXPRESSION",
       input: string };

+type NavigateAction = { type: "NAVIGATE", debuggeeUrl: string }
+
 /**
  * Actions: Source, Breakpoint, and Navigation
  *
@@ -159,5 +161,5 @@ export type Action =
   SourceAction
   | BreakpointAction
   | PauseAction
-  | { type: "NAVIGATE" }
+  | NavigateAction
   | UIAction;
```

File: src/actions/types.js [hunk 3]
```diff
@@ -159,5 +161,5 @@ export type Action =
   SourceAction
   | BreakpointAction
   | PauseAction
-  | { type: "NAVIGATE" }
+  | NavigateAction
```
Comment (wldcordeiro) [inline src/actions/types.js]: 👍

File: src/utils/sources-tree.js [hunk 4]
```diff
@@ -163,7 +163,7 @@ function isDirectory(url: Object) {
  * @memberof utils/sources-tree
  * @static
  */
-function addToTree(tree: any, source: TmpSource) {
+function addToTree(tree: any, source: TmpSource, debuggeeUrl: any) {
```
Comment (wldcordeiro) [inline src/utils/sources-tree.js]: Is the URL really an any? I don't mind using it but if there's a type we could use it's preferable.

File: src/utils/sources-tree.js [hunk 5]
```diff
@@ -224,13 +225,27 @@ function addToTree(tree: any, source: TmpSource) {
 }
 
 /**
+ * @memberof utils/sources-tree
+ * @static
+ */
+function isExactUrlMatch(pathPart:string, debuggeeUrl:string) {
```
Comment (wldcordeiro) [inline src/utils/sources-tree.js]: Testing this utility would be nice.

File: src/utils/sources-tree.js [hunk 6]
```diff
@@ -224,13 +225,27 @@ function addToTree(tree: any, source: TmpSource) {
...
+
+/**
  * Look at the nodes in the source tree, and determine the index of where to
  * insert a new node. The ordering is index -> folder -> file.
  * @memberof utils/sources-tree
  * @static
  */
 function determineFileSortOrder(nodes:Array<Node>, pathPart:string,
-                                isLastPart:boolean) {
+                                isLastPart:boolean, debuggeeUrl:any) {
```
Comment (wldcordeiro) [inline src/utils/sources-tree.js]: Same thing here, can we use something other than any?

File: src/components/SourcesTree.js [hunk 7]
```diff
@@ -24,21 +26,23 @@ type CreateTree = {
...
-    selectedSource: ImPropTypes.map
+    selectedSource: ImPropTypes.map,
+    debuggeeUrl: PropTypes.string
   },
 
   displayName: "SourcesTree",
 
   getInitialState(): CreateTree {
-    return createTree(this.props.sources);
+    return createTree(this.props.sources, this.props.debuggeeUrl || "");
```
Comment (jasonLaster) [inline src/components/SourcesTree.js]: we can improve this with this nifty trick

```js
getDefaultProps() {}
```

File: src/components/SourcesTree.js [hunk 8]
```diff
@@ -95,7 +101,7 @@ let SourcesTree = React.createClass({
 
     const uncollapsedTree = this.state.uncollapsedTree;
     for (let source of newSet) {
-      addToTree(uncollapsedTree, source);
+      addToTree(uncollapsedTree, source, this.state.debuggeeUrl);
```
Comment (jasonLaster) [inline src/components/SourcesTree.js]: shouldn't this be props?

File: src/components/SourcesTree.js [hunk 9]
```diff
@@ -237,7 +243,8 @@ module.exports = connect(
   state => {
     return {
       shownSource: getShownSource(state),
-      selectedSource: getSelectedSource(state)
+      selectedSource: getSelectedSource(state),
+      debuggeeUrl: getDebuggeeUrl(state)
```
Comment (jasonLaster) [inline src/components/SourcesTree.js]: i see... lets stick to props and just compute this every time we need it. Also - is there a better name for the function to differentiate from the raw value

File: src/utils/sources-tree.js [hunk 10]
```diff
@@ -224,13 +225,27 @@ function addToTree(tree: any, source: TmpSource) {
 }
 
 /**
+ * @memberof utils/sources-tree
+ * @static
+ */
+function isExactUrlMatch(pathPart:string, debuggeeUrl:string) {
+  // compare to hostname with an optional 'www.' prefix
+  const urlParts = /^(?:\w+\:\/\/)?([^\/]+)(.*)$/.exec(debuggeeUrl);
+  if (urlParts == null) {
```
Comment (jasonLaster) [inline src/utils/sources-tree.js]: maybe three equals `===`

File: src/utils/sources-tree.js [hunk 11]
```diff
@@ -224,13 +225,27 @@ function addToTree(tree: any, source: TmpSource) {
...
+    return false;
+  }
+  const baseUrl = urlParts[1].replace(/^www\./, "");
+  return pathPart === baseUrl || pathPart === `www.${baseUrl}`;
+}
```
Comment (jasonLaster) [inline src/utils/sources-tree.js]: what's this trying to solve?

File: src/utils/sources-tree.js [hunk 12]
```diff
@@ -242,6 +257,21 @@ function determineFileSortOrder(nodes:Array<Node>, pathPart:string,
...
+      const nodeUrlMatch = isExactUrlMatch(node.name, debuggeeUrl);
+      if (rootUrlMatch) {
+        // pathPart matches root url and must go first
+        return true;
+      }
+      if (nodeUrlMatch) {
+        // Examined item matches root url and must go first
+        return false;
+      }
+      // If neither is the case, continue to compare alphabetically
```
Comment (jasonLaster) [inline src/utils/sources-tree.js]: comments really help here

File: src/utils/sources-tree.js (hunk 11 above)
Comment (jasonLaster) [inline src/utils/sources-tree.js]: actually, i'm pretty sure there must be something in:

`URL.parse` or `new URL` that would give you a hostname you could compare

Comment (Dalimil) [inline src/utils/sources-tree.js]: I know but I cannot pass any non-browser unit-tests otherwise...

File: src/components/SourcesTree.js (hunk 7 above)
Comment (Dalimil) [inline src/components/SourcesTree.js]: hmmm... I tried it just now but flow is still complaining about a potential `undefined` value

File: src/utils/sources-tree.js (hunk 11 above)
Comment (jasonLaster) [inline src/utils/sources-tree.js]: ahh, i see - i think we solved this somewhere else as well :)

File: src/components/SourcesTree.js (hunk 7 above)
Comment (jasonLaster) [inline src/components/SourcesTree.js]: oh yeah - i'm not surprised... I'll look quickly

File: src/utils/sources-tree.js (hunk 11 above)
Comment (jasonLaster) [inline src/utils/sources-tree.js]: Hmm - we do this `const { parse } = require("url")` in several places

File: src/components/SourcesTree.js (hunk 7 above)
Comment (jasonLaster) [inline src/components/SourcesTree.js]: try adding `isRequired` to the prop with `getDefaultProps`

File: src/utils/sources-tree.js [hunk 13]
```diff
@@ -224,13 +225,27 @@ function addToTree(tree: any, source: TmpSource) {
...
+ */
+function isExactUrlMatch(pathPart:string, debuggeeUrl:string) {
+  // compare to hostname with an optional 'www.' prefix
+  const { host } = parse(debuggeeUrl);
+  if (!host) {
+    return false;
+  }
+  const baseUrl = host.replace(/^www\./, "");
+  return pathPart === baseUrl || pathPart === `www.${baseUrl}`;
+}
```
Comment (jasonLaster) [inline src/utils/sources-tree.js]: Hey @Dalimil can we try `parse` which is required in the file

Comment (Dalimil) [inline src/utils/sources-tree.js]: You mean `parse(pathPart)` ? I'm already using it for the debuggeeUrl...

Comment (jasonLaster) [inline src/utils/sources-tree.js]: oh nice - sorry missed that :)

File: src/utils/sources-tree.js [hunk 14]
```diff
@@ -224,13 +225,27 @@ function addToTree(tree: any, source: TmpSource) {
...
+ * @static
+ */
+function isExactUrlMatch(pathPart:string, debuggeeUrl:string) {
+  // compare to hostname with an optional 'www.' prefix
+  const { host } = parse(debuggeeUrl);
+  if (!host) {
+    return false;
+  }
+  const baseUrl = host.replace(/^www\./, "");
+  return pathPart === baseUrl || pathPart === `www.${baseUrl}`;
```
Comment (jasonLaster) [inline src/utils/sources-tree.js]: this seems slightly more complicated than it needs to be. Could we also replace the `www` for the pathPart?

```js
function isExactUrlMatch(pathPart:string, debuggeeUrl:string) {
  function getHost(url) {
    const { host } = parse(debuggeeUrl);
    if (!host) {
      return ""
    }
    return host.replace(/^www\./, "");
  }

  return getHost(pathPart) == getHost(debuggeeUrl)
}
```

**Comment by Dalimil** [comment]
When I try to `parse(pathPart)` I'd get a `null` for the host because the URL parsing expects a protocol but `pathPart` is usually just `www.example.com`

**Comment by jasonLaster** [comment]
ahh - i see. In that case, we can do something simpler :)

**Pull request merged**